### PR TITLE
rubocops/lines: recommend on_os/OS.os? based on context.

### PR DESCRIPTION
**Pull request by MikeMcQuaid**
Recommend the use of `on_macos` and `on_linux` unless we're in `def install`, `def post_install` or `test do` in which case recommend `OS.mac?` and `OS.linux?` instead.

TODO: 
- [x] add autocorrection, tests
- [x] fix homebrew/core

Fixes https://github.com/Homebrew/brew/issues/11687

### CODE REVIEW

File: Library/Homebrew/rubocops/lines.rb [hunk 1]
```diff
@@ -372,11 +372,50 @@ def audit_formula(_node, _class_node, _parent_class_node, body_node)
             end
           end
 
-          [:mac?, :linux?].each do |method_name|
-            next if formula_tap != "homebrew-core" || file_path&.include?("linuxbrew")
+          no_on_os_method_names = [:install, :post_install].freeze
+          no_on_os_block_names = [:test].freeze
+          [[:on_macos, :mac?], [:linux, :linux?]].each do |on_method_name, if_method_name|
```
Comment (Bo98) [inline Library/Homebrew/rubocops/lines.rb:377]:
```suggestion
          [[:on_macos, :mac?], [:on_linux, :linux?]].each do |on_method_name, if_method_name|
```

**Comment by SMillerDev** [comment]
Two requests here. I think what made the `on_os` blocks easy to read was that they _had_ to be blocks. If we can, I'd like to enforce that here too, maybe with the exception of ternary operators.

And a second, less important, request: I'd like to avoid `unless` in core since it's very ruby specific and makes formula harder to understand for the general public.

**Comment by carlocab** [comment]
Some background discussion: #11687

> I think what made the `on_os` blocks easy to read was that they _had_ to be blocks.

It depends on where they're used, in my opinion. They're nice for expressing dependencies (i.e. `depends_on`), and for fencing operations that need to be done on specific platforms. They're really bad for conditionally defining variables because they lead to patterns like

```ruby
foo = nil

on_macos do
  foo = "foo"
end

on_linux do
  foo = "bar"
end
```

or

```ruby
foo = "foo"
on_linux do
  foo = "bar"
end
```

This is because of how variable scoping works inside blocks. I guess allowing them for ternary operators or ternary-operator-like expressions would help, but that seems a lot harder to enforce than "allow `if` statements in `install`, `test`, and `post_install`, and blocks everywhere else.

I had previously contemplated turning `on_macos` and `on_linux` into methods that optionally take blocks, and return booleans otherwise as a way to avoid being forced into the patterns above (along with possibly defining aliases `on_macos?` and `on_linux?`). However, that doesn't get at what you (@SMillerDev) want here either. I suspect doing this makes them harder to standardise with Rubocop too.

CC @Homebrew/core for thoughts.

**Comment by SMillerDev** [comment]
My fear in this change is mostly going to be things like the following that IMHO would hurt readability of formulae.

```rb
./configure
system "make", "Darwin" if OS.mac?
system "make", "linux" if OS.linux?
```

Requiring blocks seemed like the best way to avoid that. If people have other ideas I'm sure I can be persuaded though.

**Comment by Rylan12** [comment]
The benefit to using a block-style `if` is that we can get the best of both worlds, right? Meaning, the easier-to-read block syntax without the problems described above. For example:

```ruby
foo = if OS.mac?
  "foo"
else
  "bar"
end
```

Did I misunderstand the concern?

**Comment by MikeMcQuaid** [comment]
> Two requests here. I think what made the `on_os` blocks easy to read was that they _had_ to be blocks. If we can, I'd like to enforce that here too, maybe with the exception of ternary operators.
> 
> And a second, less important, request: I'd like to avoid `unless` in core since it's very ruby specific and makes formula harder to understand for the general public.

I’m fine with both of these as guidance. I’d like to punt on making them required for this PR to be merged, though (because it’s going to delay it further).

**Comment by MikeMcQuaid** [comment]
> My fear in this change is mostly going to be things like the following that IMHO would hurt readability of formulae.
> 
> ```ruby
> system "./configure"
> system "make", "Darwin" if OS.mac?
> system "make", "linux" if OS.linux?
> ```
> 
> Requiring blocks seemed like the best way to avoid that. If people have other ideas I'm sure I can be persuaded though.

I think blocks didn’t look nice here either. Again a future extension for this case would be to require use of an if/else here; we shouldn’t be allowing or assuming a case which isn’t Mac or Linux.

**Review by Rylan12 — APPROVED** [review]
Looks good to me. Some tests would be good here (and probably not too difficult to write)

**Comment by MikeMcQuaid** [comment]
> Some tests would be good here

Agreed, WIP locally.



> and probably not too difficult to write

Disagreed, WIP locally 😂.

**Comment by Rylan12** [comment]
> > and probably not too difficult to write
> 
> Disagreed, WIP locally 😂.

😭 Which part is giving you trouble? I thought it would be a simple rubocop `expect_offense`/`expect_correction` kind of thing?

**Review by carlocab — APPROVED** [review]
Thanks Mike. ❤️ 

Happy to consider #11688 resolved when this is merged.

**Comment by MikeMcQuaid** [comment]
Thanks for all the help here!

**Pull request merged**